### PR TITLE
Permit `asm_const` and `asm_sym` to reference generic params

### PR DESCRIPTION
**Pull request by nbdd0121**
Related #96557

These constructs will be allowed:
```rust
fn foofoo<const N: usize>() {}

unsafe fn foo<const N: usize>() {
    asm!("/* {0} */", const N);
    asm!("/* {0} */", const N + 1);
    asm!("/* {0} */", sym foofoo::<N>);
}

fn barbar<T>() {}

unsafe fn bar<T>() {
    asm!("/* {0} */", const std::mem::size_of::<T>());
    asm!("/* {0} */", const std::mem::size_of::<(T, T)>());
    asm!("/* {0} */", sym barbar::<T>);
    asm!("/* {0} */", sym barbar::<(T, T)>);
}
```

@Amanieu, I didn't switch inline asms to use `DefKind::InlineAsm`, as I see little value doing that; given that no type inference is needed, it will only make typecking slower and more complex but will have no real gains. I did switch them to follow the same code path as inline asm during symbol resolution, though.
The `error: unconstrained generic constant` you mentioned in #76001 is due to the fact that `to_const` will actually add a wfness obligation to the constant, which we don't need for `asm_const`, so I have that removed.

@rustbot label: +A-inline-assembly +F-asm

### CODE REVIEW

**Comment by Amanieu** [comment]
LGTM!

File: src/test/ui/asm/generic-const.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    asm!("/* {0} */", sym foofoo::<N>);
+}
+
+fn barbar<T>() {}
+
+unsafe fn bar<T>() {
+    asm!("/* {0} */", const std::mem::size_of::<T>());
+    asm!("/* {0} */", const std::mem::size_of::<(T, T)>());
+    asm!("/* {0} */", sym barbar::<T>);
+    asm!("/* {0} */", sym barbar::<(T, T)>);
```
Comment (nagisa) [inline src/test/ui/asm/generic-const.rs:22]: We are missing a feature gate test for these constructs with generics I think.

**Comment by nagisa** [comment]
r=me once asm-const and asm-sym feature gate tests are augmented.

**Comment by Amanieu** [comment]
`sym`  and `const` operands are still unstable, is it necessary to add a separate feature gate for their use with generics?

**Comment by nagisa** [comment]
I don't think its necessary to add an entirely separate test, but I think augmenting the existing `feature-gate-asm_sym` and `feature-gate-asm_const` with just one additional line each is probably worthwhile, if only to serve as a reminder when we look back at stabilizing these features.

**Pull request merged**